include/boost/geometry/util/for_each_dimension.hpp

```diff
@@ -0,0 +1,409 @@
+struct for_each_index_impl1<F, S, F>
+{
+    template <typename UnaryPredicate>
+    constexpr static inline void apply(UnaryPredicate& )
+    {}
+};
+
+// O(logN) version 2
+
+template <std::size_t N>
+struct for_each_index_impl2
+{
+    static const std::size_t N1 = N / 2;
+    static const std::size_t N2 = N - N1;
+
+    template <std::size_t Offset, typename UnaryPredicate>
+    constexpr static inline void apply(UnaryPredicate& predicate)
+    {
+        for_each_index_impl2<N1>::template apply<Offset>(predicate);
+        for_each_index_impl2<N2>::template apply<Offset + N1>(predicate);
+    }
+};
+
+template <>
+struct for_each_index_impl2<3>
+{
+    template <std::size_t Offset, typename UnaryPredicate>
+    constexpr static inline void apply(UnaryPredicate& predicate)
+    {
+        predicate(util::index_constant<Offset>());
+        predicate(util::index_constant<Offset + 1>());
+        predicate(util::index_constant<Offset + 2>());
+    }
+};
+
+template <>
+struct for_each_index_impl2<2>
+{
+    template <std::size_t Offset, typename UnaryPredicate>
+    constexpr static inline void apply(UnaryPredicate& predicate)
+    {
+        predicate(util::index_constant<Offset>());
+        predicate(util::index_constant<Offset + 1>());
+    }
+};
+
+template <>
+struct for_each_index_impl2<1>
+{
+    template <std::size_t Offset, typename UnaryPredicate>
+    constexpr static inline void apply(UnaryPredicate& predicate)
+    {
+        predicate(util::index_constant<Offset>());
+    }
+};
+
+template <>
+struct for_each_index_impl2<0>
+{
+    template <std::size_t Offset, typename UnaryPredicate>
+    constexpr static inline void apply(UnaryPredicate& )
+    {}
+};
+
+// C++17 fold expression
+
+#ifndef BOOST_NO_CXX17_FOLD_EXPRESSIONS
+
+template <typename UnaryPredicate, std::size_t ...Is>
+constexpr inline UnaryPredicate for_each_dimension_impl(UnaryPredicate& predicate,
+                                                        std::index_sequence<Is...>)
+{
+    (predicate(util::index_constant<Is>()), ...);
+    return predicate;
+}
+
+#endif
+
+// Interface
+
+template <std::size_t N, typename UnaryPredicate>
+constexpr inline UnaryPredicate for_each_index(UnaryPredicate predicate)
+{
+    for_each_index_impl
+        <
+            0, N
+        >::apply(predicate);
+    return predicate;
+}
+
+template <typename Geometry, typename UnaryPredicate>
+constexpr inline UnaryPredicate for_each_dimension(UnaryPredicate predicate)
+{
+//#ifndef BOOST_NO_CXX17_FOLD_EXPRESSIONS
+//    for_each_dimension_impl(predicate,
+//        std::make_index_sequence<geometry::dimension<Geometry>::value>());
+//#else
+    for_each_index_impl2
+        <
+            geometry::dimension<Geometry>::value
+        >::template apply<0>(predicate);
+//#endif
+    return predicate;
+}
+
+// ----------------------------------------------------------------------------
+
+// O(logN) version 2
+
+template <std::size_t N>
+struct all_indexes_of_impl2
```

Comment:
Maybe you can split this file into the various algorithms?

Reply:
There are 8 algorithms there so technically they could be divided into 2, 4 or 8 files. The corresponding STL algorithms are defined in `<algorithm>`. So what do you think about renaming the file to `util/algorithm.hpp` and keeping them together?

Btw, what do you think about the namespace? Currently they are in `bg::detail` which gives us the most flexibility. But technically they could be moved to `bg::util` or simply `bg`.

Reply:
Sorry, I missed your answer.
Right, it makes sense, you can keep them then and change the filename.
I think they're useful also for end users, so if you think they are in final form then `bg` is fine for me. If there is any doubt then I would indeed put them for now in `bg::detail`

Reply:
Done. I kept it in the `detail` because it allows us to change it more easily. We can always move it if a user asks for it.